jerry-core/ecma/base/ecma-globals.h

```diff
@@ -204,20 +204,99 @@ typedef enum
+ *   The type of each item is stored in the first byte of the item.
+ *
+ *   The most common item is the property pair, which contains two
+ *   ecmascript properties. It is also important, that after the
+ *   first property pair, only property pair items are allowed.
+ *
+ *   Example for other items is property name hash map, or array of items.
+ */
+
+/**
+ * Property type list.
  */
 typedef enum
 {
-  ECMA_PROPERTY_FLAG_NAMEDDATA = 1u << 0, /**< property is named data */
-  ECMA_PROPERTY_FLAG_NAMEDACCESSOR = 1u << 1, /**< property is named accessor */
-  ECMA_PROPERTY_FLAG_INTERNAL = 1u << 2, /**< property is internal property */
-  ECMA_PROPERTY_FLAG_CONFIGURABLE = 1u << 3, /**< property is configurable */
-  ECMA_PROPERTY_FLAG_ENUMERABLE = 1u << 4, /**< property is enumerable */
-  ECMA_PROPERTY_FLAG_WRITABLE = 1u << 5, /**< property is writable */
-  ECMA_PROPERTY_FLAG_LCACHED = 1u << 6, /**< property is lcached */
+  ECMA_PROPERTY_TYPE_DELETED, /**< deleted property */
+  ECMA_PROPERTY_TYPE_INTERNAL, /**< internal property */
+  ECMA_PROPERTY_TYPE_NAMEDDATA, /**< property is named data */
+  ECMA_PROPERTY_TYPE_NAMEDACCESSOR, /**< property is named accessor */
+
+  ECMA_PROPERTY_TYPE_HASH, /**< property hash (TODO: implement this) */
+
+  ECMA_PROPERTY_TYPE_PROPERTY_PAIR__MAX = ECMA_PROPERTY_TYPE_NAMEDACCESSOR, /**< highest value for
+                                                                             *   property pair types. */
+  ECMA_PROPERTY_TYPE__MAX = ECMA_PROPERTY_TYPE_HASH, /**< highest value for property types. */
+} ecma_property_types_t;
+
+/**
+ * Property type mask.
+ */
+#define ECMA_PROPERTY_TYPE_MASK 0x7
+
+/**
+ * Property flags base shift.
+ */
+#define ECMA_PROPERTY_FLAG_SHIFT 3
+
+/**
+ * Property flag list (for ECMA_PROPERTY_TYPE_NAMEDDATA
+ * and ECMA_PROPERTY_TYPE_NAMEDACCESSOR).
+ */
+typedef enum
+{
+  ECMA_PROPERTY_FLAG_CONFIGURABLE = 1u << (ECMA_PROPERTY_FLAG_SHIFT + 0), /**< property is configurable */
+  ECMA_PROPERTY_FLAG_ENUMERABLE = 1u << (ECMA_PROPERTY_FLAG_SHIFT + 1), /**< property is enumerable */
+  ECMA_PROPERTY_FLAG_WRITABLE = 1u << (ECMA_PROPERTY_FLAG_SHIFT + 2), /**< property is writable */
+  ECMA_PROPERTY_FLAG_LCACHED = 1u << (ECMA_PROPERTY_FLAG_SHIFT + 3), /**< property is lcached */
 } ecma_property_flags_t;
 
 /**
+ * Abstract property representation.
+ *
+ * A property is a type_and_flags byte and an ecma_value_t value pair.
+ * This pair is represented by a single pointer in JerryScript. Although
+ * a packed struct would only consume sizeof(ecma_value_t)+1 memory
+ * bytes, accessing such structure is inefficient from the CPU viewpoint
+ * because the value is not naturally aligned. To improve performance,
+ * type bytes and values are packed together. It ecma_value_t is four
+ * bytes long, maximum four properties can be packed together:
+ *
+ *  [Type 1][Type 2][Type 3][Type 4][Value 1][Value 2][Value 3][Value 4]
```

Comment:
This explanation can be a bit misleading. How many ecma values are used in the above example? My guess is 5. If I'm right, then I'd suggest writing the above example as `[Type 1,Type 2,Type 3,Type 4][Value 1][Value 2][Value 3][Value 4]` so that the brackets always denote `ecma_value_t` boundaries.

(If I'm wrong, then we need an even better explanation...)